In [3]:
! pip install geopandas
! pip install pysal

## Import Necessary Packages

In [12]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import geopandas as gpd
import pysal as ps
import shapely.geometry
from pyproj import CRS

from pandas import DataFrame,Series

## What is Pandas?

- Pandas provides a SQL-like approach (that blends in elements of statistics and linear algebra) to analyzing tables of data
- DataFrames in R are very similar
- Pandas has been adopted as a de facto standard for input and vectorization across numerous disciplines including Python data analysis with spatial components

### Loading data from a csv

In [5]:
basedf=pd.read_csv('ss13hil.csv')
basedf[['SERIALNO', 'PUMA00', 'PUMA10', 'ST', 'ADJHSG', 'ADJINC',
        'WGTP', 'NP', 'TYPE', 'ACR', 'AGS', 'BATH', 'BDSP',
        'BLD', 'BUS', 'CONP', 'ELEP', 'FS', 'FULP']].head()

SERIALNO  PUMA00  PUMA10  ST   ADJHSG   ADJINC  WGTP  NP  TYPE  ACR  \
0  2009000000061    3515      -9  17  1086032  1085467    36   0     1  NaN   
1  2009000000075    1000      -9  17  1086032  1085467     6   1     1  1.0   
2  2009000000108    3402      -9  17  1086032  1085467    15   3     1  1.0   
3  2009000000132    3510      -9  17  1086032  1085467    60   4     1  1.0   
4  2009000000150    3518      -9  17  1086032  1085467    37   3     1  1.0   

   AGS  BATH  BDSP  BLD  BUS  CONP   ELEP   FS  FULP  
0  NaN   2.0   2.0  8.0  NaN   0.0    NaN  NaN   NaN  
1  NaN   1.0   3.0  1.0  2.0   0.0  200.0  2.0   2.0  
2  NaN   1.0   3.0  2.0  2.0   0.0   80.0  2.0   2.0  
3  NaN   1.0   3.0  2.0  2.0   0.0    1.0  2.0   2.0  
4  NaN   1.0   3.0  2.0  2.0   0.0  200.0  1.0   2.0

In [27]:
basedf=pd.read_csv('ss13hil.csv', index_col='SERIALNO',
                   usecols=['SERIALNO', 'PUMA00', 'PUMA10', 'ST',
                            'ADJHSG', 'ADJINC', 'WGTP', 'NP', 'TYPE', 'ACR',
                            'AGS', 'BATH', 'BDSP', 'BLD', 'BUS', 'CONP', 'ELEP',
                            'FS', 'FULP'])
basedf.head()

PUMA00  PUMA10  ST   ADJHSG   ADJINC  WGTP  NP  TYPE  ACR  AGS  \
SERIALNO                                                                        
2009000000061    3515      -9  17  1086032  1085467    36   0     1  NaN  NaN   
2009000000075    1000      -9  17  1086032  1085467     6   1     1  1.0  NaN   
2009000000108    3402      -9  17  1086032  1085467    15   3     1  1.0  NaN   
2009000000132    3510      -9  17  1086032  1085467    60   4     1  1.0  NaN   
2009000000150    3518      -9  17  1086032  1085467    37   3     1  1.0  NaN   

               BATH  BDSP  BLD  BUS  CONP   ELEP   FS  FULP  
SERIALNO                                                     
2009000000061   2.0   2.0  8.0  NaN   0.0    NaN  NaN   NaN  
2009000000075   1.0   3.0  1.0  2.0   0.0  200.0  2.0   2.0  
2009000000108   1.0   3.0  2.0  2.0   0.0   80.0  2.0   2.0  
2009000000132   1.0   3.0  2.0  2.0   0.0    1.0  2.0   2.0  
2009000000150   1.0   3.0  2.0  2.0   0.0  200.0  1.0   2.0

### query
- query takes a text string argument in the form (roughly) of a SQL WHERE clause
- Column names need to be referenced without quoting so suitable single-word names are needed
- https://pandas.pydata.org/pandas-docs/version/0.22/indexing.html#indexing-query

In [7]:
basedf.query('PUMA00==3515')

PUMA00  PUMA10  ST   ADJHSG   ADJINC  WGTP  NP  TYPE  ACR  AGS  \
SERIALNO                                                                        
2009000000061    3515      -9  17  1086032  1085467    36   0     1  NaN  NaN   
2009000002489    3515      -9  17  1086032  1085467    15   1     1  2.0  1.0   
2009000002611    3515      -9  17  1086032  1085467    49   2     1  NaN  NaN   
2009000002724    3515      -9  17  1086032  1085467    45   1     1  1.0  NaN   
2009000006025    3515      -9  17  1086032  1085467    17   2     1  NaN  NaN   
2009000009853    3515      -9  17  1086032  1085467    14   4     1  1.0  NaN   
2009000010773    3515      -9  17  1086032  1085467    18   2     1  1.0  NaN   
2009000012599    3515      -9  17  1086032  1085467    60   4     1  1.0  NaN   
2009000012695    3515      -9  17  1086032  1085467    44   3     1  1.0  NaN   
2009000013780    3515      -9  17  1086032  1085467    20   3     1  NaN  NaN   
2009000021230    3515      -9  17  1086032  1085467    18   1     1  NaN  NaN   
2009000022438    3515      -9  17  1086032  1085467     0   1     2  NaN  NaN   
2009000022715    3515      -9  17  1086032  1085467    33   1     1  NaN  NaN   
2009000022871    3515      -9  17  1086032  1085467    27   7     1  1.0  NaN   
2009000023991    3515      -9  17  1086032  1085467    34   1     1  NaN  NaN   
2009000024254    3515      -9  17  1086032  1085467    15   4     1  NaN  NaN   
2009000026127    3515      -9  17  1086032  1085467    24   1     1  NaN  NaN   
2009000026822    3515      -9  17  1086032  1085467    34   0     1  NaN  NaN   
2009000027183    3515      -9  17  1086032  1085467    23   2     1  NaN  NaN   
2009000027392    3515      -9  17  1086032  1085467    37   0     1  NaN  NaN   
2009000028444    3515      -9  17  1086032  1085467    21   1     1  NaN  NaN   
2009000028568    3515      -9  17  1086032  1085467    14   0     1  1.0  NaN   
2009000030494    3515      -9  17  1086032  1085467    39   6     1  NaN  NaN   
2009000031347    3515      -9  17  1086032  1085467    31   1     1  NaN  NaN   
2009000032751    3515      -9  17  1086032  1085467     0   1     2  NaN  NaN   
2009000032821    3515      -9  17  1086032  1085467    40   0     1  NaN  NaN   
2009000039391    3515      -9  17  1086032  1085467    41   1     1  NaN  NaN   
2009000039418    3515      -9  17  1086032  1085467    17   1     1  NaN  NaN   
2009000042011    3515      -9  17  1086032  1085467    33   3     1  NaN  NaN   
2009000046340    3515      -9  17  1086032  1085467    40   4     1  NaN  NaN   
...               ...     ...  ..      ...      ...   ...  ..   ...  ...  ...   
2011001440178    3515      -9  17  1035725  1054614     0   1     3  NaN  NaN   
2011001443385    3515      -9  17  1035725  1054614    11   1     1  NaN  NaN   
2011001443607    3515      -9  17  1035725  1054614     0   1     2  NaN  NaN   
2011001451083    3515      -9  17  1035725  1054614     8   1     1  1.0  NaN   
2011001451956    3515      -9  17  1035725  1054614    38   6     1  NaN  NaN   
2011001452314    3515      -9  17  1035725  1054614    63   1     1  NaN  NaN   
2011001452423    3515      -9  17  1035725  1054614    17   1     1  NaN  NaN   
2011001455146    3515      -9  17  1035725  1054614    21   1     1  NaN  NaN   
2011001455997    3515      -9  17  1035725  1054614     0   1     2  NaN  NaN   
2011001458835    3515      -9  17  1035725  1054614    14   2     1  NaN  NaN   
2011001459951    3515      -9  17  1035725  1054614    31   2     1  NaN  NaN   
2011001462339    3515      -9  17  1035725  1054614    11   4     1  NaN  NaN   
2011001462348    3515      -9  17  1035725  1054614    27   6     1  1.0  NaN   
2011001462403    3515      -9  17  1035725  1054614    31   2     1  1.0  NaN   
2011001462832    3515      -9  17  1035725  1054614    23   0     1  NaN  NaN   
2011001465772    3515      -9  17  1035725  1054614     6   1     1  NaN  NaN   
2011001472550    3515      -9  17  1035725  10

- The query functionality can work between fields.
- However, the only operators that I would rely on are (==, !=, <, >, <=, >=, &, | )
- query() is by default evaluated using the numexpr engine, which outperforms pure python on DataFrames of more than 200,000 rows

In [8]:
basedf.query('BDSP==NP')

PUMA00  PUMA10  ST   ADJHSG   ADJINC  WGTP  NP  TYPE  ACR  AGS  \
SERIALNO                                                                        
2009000000108    3402      -9  17  1086032  1085467    15   3     1  1.0  NaN   
2009000000150    3518      -9  17  1086032  1085467    37   3     1  1.0  NaN   
2009000000225    3402      -9  17  1086032  1085467    19   2     1  NaN  NaN   
2009000000256     600      -9  17  1086032  1085467    26   2     1  NaN  NaN   
2009000000335    2700      -9  17  1086032  1085467    28   1     1  NaN  NaN   
2009000000461     400      -9  17  1086032  1085467    43   4     1  1.0  NaN   
2009000000496    1101      -9  17  1086032  1085467    53   2     1  NaN  NaN   
2009000000629    3001      -9  17  1086032  1085467    22   3     1  1.0  NaN   
2009000000775    3510      -9  17  1086032  1085467    24   1     1  NaN  NaN   
2009000000912    1800      -9  17  1086032  1085467     7   1     1  1.0  NaN   
2009000001130    3005      -9  17  1086032  1085467    19   1     1  NaN  NaN   
2009000001170    1000      -9  17  1086032  1085467    22   3     1  2.0  1.0   
2009000001233     200      -9  17  1086032  1085467     8   1     1  1.0  NaN   
2009000001398    3305      -9  17  1086032  1085467    13   4     1  1.0  NaN   
2009000001474    3104      -9  17  1086032  1085467    21   5     1  1.0  NaN   
2009000001515    3502      -9  17  1086032  1085467    13   2     1  NaN  NaN   
2009000001591     200      -9  17  1086032  1085467     2   4     1  1.0  NaN   
2009000001736    2700      -9  17  1086032  1085467    19   4     1  1.0  NaN   
2009000001902    3206      -9  17  1086032  1085467    17   3     1  1.0  NaN   
2009000001950    2900      -9  17  1086032  1085467    16   3     1  1.0  NaN   
2009000002087     500      -9  17  1086032  1085467     3   3     1  1.0  NaN   
2009000002113    1300      -9  17  1086032  1085467    20   2     1  2.0  1.0   
2009000002240     900      -9  17  1086032  1085467    24   2     1  NaN  NaN   
2009000002293    3510      -9  17  1086032  1085467    23   1     1  NaN  NaN   
2009000002317    3405      -9  17  1086032  1085467    12   4     1  1.0  NaN   
2009000002429    3410      -9  17  1086032  1085467    24   4     1  1.0  NaN   
2009000002483    3511      -9  17  1086032  1085467    13   2     1  NaN  NaN   
2009000002496    3206      -9  17  1086032  1085467    17   4     1  1.0  NaN   
2009000002722    1400      -9  17  1086032  1085467    14   3     1  1.0  NaN   
2009000002726    3204      -9  17  1086032  1085467    20   3     1  1.0  NaN   
...               ...     ...  ..      ...      ...   ...  ..   ...  ...  ...   
2013001488955      -9    2000  17  1000000  1007549    20   2     1  NaN  NaN   
2013001488994      -9    3521  17  1000000  1007549    15   2     1  NaN  NaN   
2013001489051      -9    2501  17  1000000  1007549     4   2     1  1.0  NaN   
2013001489062      -9    3007  17  1000000  1007549    41   2     1  1.0  NaN   
2013001489147      -9    3420  17  1000000  1007549    77   3     1  NaN  NaN   
2013001489260      -9    3502  17  1000000  1007549    11   1     1  NaN  NaN   
2013001489329      -9    2100  17  1000000  1007549     5   1     1  1.0  NaN   
2013001489749      -9    3102  17  1000000  1007549    41   1     1  NaN  NaN   
2013001489860      -9    1300  17  1000000  1007549    64   5     1  1.0  NaN   
2013001489873      -9    3205  17  1000000  1007549    49   1     1  NaN  NaN   
2013001489894      -9    3204  17  1000000  1007549    27   4     1  1.0  NaN   
2013001489972      -9    3700  17  1000000  1007549     3   2     1  1.0  NaN   
2013001489989      -9     105  17  1000000  1007549    32   1     1  1.0  NaN   
2013001490108      -9    3530  17  1000000  1007549    17   3     1  1.0  NaN   
2013001490119      -9    1701  17  1000000  1007549    17   3     1  1.0  NaN   
2013001490258      -9     300  17  1000000  1007549    13   3     1  2.0  3.0   
2013001490268      -9     501  17  1000000  10

- Arithmetic is possible although I can’t speak to its efficiency
- The use of in and not in operators as well as ==[’a’,’b’,...], although parts of this will generally be evaluated using pure python


In [10]:
basedf.query('0<BDSP<NP & PUMA00==3515')

PUMA00  PUMA10  ST   ADJHSG   ADJINC  WGTP  NP  TYPE  ACR  AGS  \
SERIALNO                                                                        
2009000002611    3515      -9  17  1086032  1085467    49   2     1  NaN  NaN   
2009000013780    3515      -9  17  1086032  1085467    20   3     1  NaN  NaN   
2009000022871    3515      -9  17  1086032  1085467    27   7     1  1.0  NaN   
2009000024254    3515      -9  17  1086032  1085467    15   4     1  NaN  NaN   
2009000030494    3515      -9  17  1086032  1085467    39   6     1  NaN  NaN   
2009000046340    3515      -9  17  1086032  1085467    40   4     1  NaN  NaN   
2009000069632    3515      -9  17  1086032  1085467    30   4     1  NaN  NaN   
2009000086634    3515      -9  17  1086032  1085467    15   4     1  1.0  NaN   
2009000088409    3515      -9  17  1086032  1085467    40   5     1  NaN  NaN   
2009000124024    3515      -9  17  1086032  1085467    25   5     1  1.0  NaN   
2009000134542    3515      -9  17  1086032  1085467    37   6     1  1.0  NaN   
2009000135687    3515      -9  17  1086032  1085467    31   5     1  NaN  NaN   
2009000137681    3515      -9  17  1086032  1085467    40   5     1  NaN  NaN   
2009000143272    3515      -9  17  1086032  1085467    55   3     1  NaN  NaN   
2009000164550    3515      -9  17  1086032  1085467    42   3     1  NaN  NaN   
2009000235371    3515      -9  17  1086032  1085467    22   4     1  NaN  NaN   
2009000242105    3515      -9  17  1086032  1085467    30   4     1  NaN  NaN   
2009000261112    3515      -9  17  1086032  1085467    48   4     1  NaN  NaN   
2009000278006    3515      -9  17  1086032  1085467    15   4     1  NaN  NaN   
2009000281193    3515      -9  17  1086032  1085467    23  13     1  1.0  NaN   
2009000289745    3515      -9  17  1086032  1085467    71   3     1  NaN  NaN   
2009000290319    3515      -9  17  1086032  1085467    30   3     1  NaN  NaN   
2009000298561    3515      -9  17  1086032  1085467    16   4     1  1.0  NaN   
2009000302575    3515      -9  17  1086032  1085467    18   3     1  NaN  NaN   
2009000304973    3515      -9  17  1086032  1085467    18   6     1  1.0  NaN   
2009000310241    3515      -9  17  1086032  1085467    49   3     1  NaN  NaN   
2009000320632    3515      -9  17  1086032  1085467    30   5     1  NaN  NaN   
2009000343682    3515      -9  17  1086032  1085467    17   3     1  NaN  NaN   
2009000345878    3515      -9  17  1086032  1085467    16   5     1  1.0  NaN   
2009000347349    3515      -9  17  1086032  1085467    10   3     1  NaN  NaN   
...               ...     ...  ..      ...      ...   ...  ..   ...  ...  ...   
2011001168857    3515      -9  17  1035725  1054614    20   2     1  NaN  NaN   
2011001169212    3515      -9  17  1035725  1054614    21   4     1  NaN  NaN   
2011001176565    3515      -9  17  1035725  1054614    19   3     1  NaN  NaN   
2011001205856    3515      -9  17  1035725  1054614    15   7     1  1.0  NaN   
2011001218196    3515      -9  17  1035725  1054614    20   3     1  NaN  NaN   
2011001218409    3515      -9  17  1035725  1054614    15   3     1  NaN  NaN   
2011001234936    3515      -9  17  1035725  1054614    25   7     1  1.0  NaN   
2011001237682    3515      -9  17  1035725  1054614    24   7     1  NaN  NaN   
2011001245795    3515      -9  17  1035725  1054614    20   3     1  NaN  NaN   
2011001249889    3515      -9  17  1035725  1054614    47   3     1  NaN  NaN   
2011001253824    3515      -9  17  1035725  1054614     8   3     1  NaN  NaN   
2011001267924    3515      -9  17  1035725  1054614    28   3     1  1.0  NaN   
2011001270004    3515      -9  17  1035725  1054614     8   2     1  NaN  NaN   
2011001273174    3515      -9  17  1035725  1054614    18   4     1  NaN  NaN   
2011001275567    3515      -9  17  1035725  1054614    19   7     1  1.0  NaN   
2011001282747    3515      -9  17  1035725  1054614     8   3     1  NaN  NaN   
2011001314614    3515      -9  17  1035725  10

### concat as JOIN
- If the indexes are overlapping and the column names are not the same and axis=1 is used, that is identical to that of INNER JOIN from SQL

In [13]:
df1=DataFrame({'a':[1,2,3], 'b':[4,5,6]}, index=['x','y','z'])
df2=DataFrame({'c':[7,8,9], 'd':[10,11,12]}, index=['x','y','z'])
pd.concat([df1,df2],axis=1)

a  b  c   d
x  1  4  7  10
y  2  5  8  11
z  3  6  9  12

### merge as JOIN
- merge is a holistic JOIN operator
- Like SQL JOINs, the options for using it are complex and take a great deal of practice to master
- We will focus on two options: on= and how=
- on determines the common column used to join the two together (a list of common columns can be specified)
- note that the indexes are not preserved. To keep them .reset index() before joining and then set the index from that column after or join on index (not covered here)

In [14]:
df1=DataFrame({'a':[1,2,3], 'b':[4,5,6]}, index=['x','y','z'])
df2=DataFrame({'a':[1,2,3], 'c':[10,11,12]}, index=['u','v','w'])
pd.merge(df1,df2,on='a')

a  b   c
0  1  4  10
1  2  5  11
2  3  6  12

### merge as INNER JOIN
- how can be set to left, right, inner, or outer
- The left on and right on options specify the matching columns on the left and right join tables if they have different names
- Note that the default value of how is inner and this will filter out non-matching rows symmetrically

In [15]:
df1=DataFrame({'a1':[1,2,3], 'b':[4,5,6]})
df2=DataFrame({'a2':[1,2,7], 'c':[10,11,12]})
pd.merge(df1,df2,how='inner',left_on='a1',right_on='a2')

a1  b  a2   c
0   1  4   1  10
1   2  5   2  11

### merge as multi-key JOIN
- By passing a list to each of the on options, the corresponding keys are matched sequentially
- In this case two rows are found to match if and only if the value of a1 matches a2 and key1 matches key2

In [16]:
df1=DataFrame({'a1':[1,2,3],'key1':['R','R','C'] ,'b':[4,5,6]})
df2=DataFrame({'a2':[1,2,7],'key2':['R','D','C'], 'c':[10,11,12]})
pd.merge(df1,df2,how='outer',left_on=['a1','key1'],right_on=['a2','key2'])

a1 key1    b   a2 key2     c
0  1.0    R  4.0  1.0    R  10.0
1  2.0    R  5.0  NaN  NaN   NaN
2  3.0    C  6.0  NaN  NaN   NaN
3  NaN  NaN  NaN  2.0    D  11.0
4  NaN  NaN  NaN  7.0    C  12.0

### map as Transforming Columns
- .map(f) applies to a Series
- It iterates efficiently over every element of the series and applies the function f to that element
- Then it assembles a new Series comprised of the transformed elements in the same order with the same index and returns it
- Weak implicit typing is critical here

In [21]:
basedf['NP\_sq']=basedf['NP'].map(lambda x: x**2)
basedf['PUMA\_str']=basedf['PUMA00'].map(lambda x: 'PUMA:'+str(x))

### Split-Apply-Combine as an overall strategy
- Similar to (but more general than) GROUP BY in SQL
- General tool for bulk changes
- The splitting step breaks data into groups using any column (including the row number)
- This can be accomplished using df.groupby('year')

### Split-Apply-Combine in more detail
- Groups produced by the split can be individually transformed by an arbitrary function
- This is the essence of Apply (the DataFrame extension of Map)
- The result is combined back into a single DataFrame

In [35]:
subdf = basedf.query('PUMA00 == 3515').copy()

def computeWeightedNP(x):
    x['weightedNP'] = x['NP']*x['WGTP']
    return x

subdf = subdf.apply(computeWeightedNP, axis=1)
totals = subdf.sum()
totals['weightedNP'] / totals['WGTP']

1.707378252712293

### apply as CROSS APPLY

In [38]:
def computeWeightedNP(x):
    x['weightedNP'] = x['NP'] * x['WGTP']
    #print(x)
    totals = x.sum()
    x['avgNP'] = totals['weightedNP'] / totals['WGTP']
    return x

subdf.groupby(['PUMA00']).apply(computeWeightedNP)

PUMA00  PUMA10    ST     ADJHSG     ADJINC  WGTP   NP  TYPE  \
SERIALNO                                                                     
2009000000061  3515.0    -9.0  17.0  1086032.0  1085467.0  36.0  0.0   1.0   
2009000002489  3515.0    -9.0  17.0  1086032.0  1085467.0  15.0  1.0   1.0   
2009000002611  3515.0    -9.0  17.0  1086032.0  1085467.0  49.0  2.0   1.0   
2009000002724  3515.0    -9.0  17.0  1086032.0  1085467.0  45.0  1.0   1.0   
2009000006025  3515.0    -9.0  17.0  1086032.0  1085467.0  17.0  2.0   1.0   
2009000009853  3515.0    -9.0  17.0  1086032.0  1085467.0  14.0  4.0   1.0   
2009000010773  3515.0    -9.0  17.0  1086032.0  1085467.0  18.0  2.0   1.0   
2009000012599  3515.0    -9.0  17.0  1086032.0  1085467.0  60.0  4.0   1.0   
2009000012695  3515.0    -9.0  17.0  1086032.0  1085467.0  44.0  3.0   1.0   
2009000013780  3515.0    -9.0  17.0  1086032.0  1085467.0  20.0  3.0   1.0   
2009000021230  3515.0    -9.0  17.0  1086032.0  1085467.0  18.0  1.0   1.0   
2009000022438  3515.0    -9.0  17.0  1086032.0  1085467.0   0.0  1.0   2.0   
2009000022715  3515.0    -9.0  17.0  1086032.0  1085467.0  33.0  1.0   1.0   
2009000022871  3515.0    -9.0  17.0  1086032.0  1085467.0  27.0  7.0   1.0   
2009000023991  3515.0    -9.0  17.0  1086032.0  1085467.0  34.0  1.0   1.0   
2009000024254  3515.0    -9.0  17.0  1086032.0  1085467.0  15.0  4.0   1.0   
2009000026127  3515.0    -9.0  17.0  1086032.0  1085467.0  24.0  1.0   1.0   
2009000026822  3515.0    -9.0  17.0  1086032.0  1085467.0  34.0  0.0   1.0   
2009000027183  3515.0    -9.0  17.0  1086032.0  1085467.0  23.0  2.0   1.0   
2009000027392  3515.0    -9.0  17.0  1086032.0  1085467.0  37.0  0.0   1.0   
2009000028444  3515.0    -9.0  17.0  1086032.0  1085467.0  21.0  1.0   1.0   
2009000028568  3515.0    -9.0  17.0  1086032.0  1085467.0  14.0  0.0   1.0   
2009000030494  3515.0    -9.0  17.0  1086032.0  1085467.0  39.0  6.0   1.0   
2009000031347  3515.0    -9.0  17.0  1086032.0  1085467.0  31.0  1.0   1.0   
2009000032751  3515.0    -9.0  17.0  1086032.0  1085467.0   0.0  1.0   2.0   
2009000032821  3515.0    -9.0  17.0  1086032.0  1085467.0  40.0  0.0   1.0   
2009000039391  3515.0    -9.0  17.0  1086032.0  1085467.0  41.0  1.0   1.0   
2009000039418  3515.0    -9.0  17.0  1086032.0  1085467.0  17.0  1.0   1.0   
2009000042011  3515.0    -9.0  17.0  1086032.0  1085467.0  33.0  3.0   1.0   
2009000046340  3515.0    -9.0  17.0  1086032.0  1085467.0  40.0  4.0   1.0   
...               ...     ...   ...        ...        ...   ...  ...   ...   
2011001440178  3515.0    -9.0  17.0  1035725.0  1054614.0   0.0  1.0   3.0   
2011001443385  3515.0    -9.0  17.0  1035725.0  1054614.0  11.0  1.0   1.0   
2011001443607  3515.0    -9.0  17.0  1035725.0  1054614.0   0.0  1.0   2.0   
2011001451083  3515.0    -9.0  17.0  1035725.0  1054614.0   8.0  1.0   1.0   
2011001451956  3515.0    -9.0  17.0  1035725.0  1054614.0  38.0  6.0   1.0   
2011001452314  3515.0    -9.0  17.0  1035725.0  1054614.0  63.0  1.0   1.0   
2011001452423  3515.0    -9.0  17.0  1035725.0  1054614.0  17.0  1.0   1.0   
2011001455146  3515.0    -9.0  17.0  1035725.0  1054614.0  21.0  1.0   1.0   
2011001455997  3515.0    -9.0  17.0  1035725.0  1054614.0   0.0  1.0   2.0   
2011001458835  3515.0    -9.0  17.0  1035725.0  1054614.0  14.0  2.0   1.0   
2011001459951  3515.0    -9.0  17.0  1035725.0  1054614.0  31.0  2.0   1.0   
2011001462339  3515.0    -9.0  17.0  1035725.0  1054614.0  11.0  4.0   1.0   
2011001462348  3515.0    -9.0  17.0  1035725.0  1054614.0  27.0  6.0   1.0   
2011001462403  3515.0    -9.0  17.0  1035725.0  1054614.0  31.0  2.0   1.0   
2011001462832  3515.0    -9.0  17.0  1035725.0  1054614.0  23.0  0.0   1.0   
2011001465772  3515.0    -9.0  17.0  1035725.0  1054614.0   6.0  1.0   1.0   
2011001472550  3515.0    -9.0  17.0  1035725.0  1054614.0  29.0  0.0   1.0   
2011001472759  3515.0    -9.0  17.0  1035725.0  1054614.0   0.0  1.0   3.0   
2011001474048  3515.0    -9.0  17.0  

### Problems 1

1) Start from an import of the PUMS csv using the columns 'SERIALNO', 'PUMA', 'BDSP', and 'NP' (with SERIALNO as the index and a DataType of str for the PUMA column using the dtype option). load this to a variable named basedf

In [98]:
basedf = pd.read_csv('psam_h17.csv', index_col='SERIALNO',
                     usecols=['SERIALNO', 'PUMA', 'NP', 'BDSP'])
basedf.head()

PUMA  NP  BDSP
SERIALNO                     
2013000000033  1300   3   3.0
2013000000037   800   1   3.0
2013000000051   105   0   1.0
2013000000079   300   1   4.0
2013000000095  1104   4   3.0

2) Use the command basedf[['PUMA']].drop duplicates() to produce a list of the unique PUMA regions.

In [99]:
unique_df = basedf[['PUMA']].drop_duplicates()
print(len(basedf))
print(len(unique_df))
print(unique_df)

292928
88
               PUMA
SERIALNO           
2013000000033  1300
2013000000037   800
2013000000051   105
2013000000079   300
2013000000095  1104
2013000000163  3102
2013000000174   700
2013000000187  3530
2013000000196  3410
2013000000207  3414
2013000000238  1900
2013000000239  3407
2013000000244  1001
2013000000251  3309
2013000000258  3521
2013000000264  2901
2013000000348  3502
2013000000349  1701
2013000000359   900
2013000000382  1204
2013000000397  3007
2013000000482  3501
2013000000517  3417
2013000000547  3413
2013000000566  3526
2013000000727  3522
2013000000996  3409
2013000001078  1205
2013000001091  3106
2013000001095  3532
...             ...
2013000003365  3525
2013000003440  2601
2013000003504  2000
2013000003590  3529
2013000003643  3412
2013000003720  3527
2013000003748  3523
2013000003797   600
2013000003895  3105
2013000003971  1500
2013000004051  2700
2013000004550  3602
2013000004738  2801
2013000004810  3005
2013000005206  3205
2013000005281  3601
2013000005

3) Using the query command, select only those rows where the PUMA region number matches one specific one that you chose to work with from the previous step. How many records are returned? (look below the readout of sample rows to see a number) Repeat this for 4 different PUMAs and compare the counts returned for each. (what is the total?) Can you rewrite this as a single query using the OR operator? make sure the resulting counts agree.

In [102]:
count_3515 = len(basedf.query('PUMA == 1300'))
count_3003 = len(basedf.query('PUMA == 800'))
count_1400 = len(basedf.query('PUMA == 3102'))
count_3502 = len(basedf.query('PUMA == 3502'))
print(count_3515)
print(count_3003)
print(count_1400)
print(count_3502)

count_sum = count_3515 + count_3003 + count_1400 + count_3502
print(count_sum)

5200
5653
2821
4633
18307


In [105]:
PUMA_check = 'PUMA == 1300 | PUMA == 800 | PUMA == 3102 | PUMA == 3502'
# combined_query = basedf.query(PUMA00_check)
# print(len(combined_query))

NP_check = 'NP > 4'

combined_check = "(" + PUMA_check + ") & " + NP_check
# print(combined_check)

combined_query = basedf.query(combined_check)
# print(combined_query)
print(len(combined_query))
combined_query.head()

959


PUMA  NP  BDSP
SERIALNO                     
2013000001418  3102   5   5.0
2013000002506  3102   5   3.0
2013000014713   800   5   1.0
2013000015250   800   5   3.0
2013000020031  1300   5   4.0

In [106]:
def PUMA_filt(x):
    if x in [1300, 800, 3102, 3502]:
        return True
    else:
        return False  
    
def NP_filt(x):
    if x > 4:
        return True
    else:
        return False
        
combined_mask = (basedf['NP'].map(NP_filt)) & (basedf['PUMA'].map(PUMA_filt))
combined_masked = basedf[combined_mask]
print(len(combined_masked))

combined_query.head()

959


PUMA  NP  BDSP
SERIALNO                     
2013000001418  3102   5   5.0
2013000002506  3102   5   3.0
2013000014713   800   5   1.0
2013000015250   800   5   3.0
2013000020031  1300   5   4.0

### Problems 2

1) Select off two subsets of the data using queries that on PUMAs. The first should include 01300, 00800, and 00105. The second should include 00105, 00300, 01104.

In [107]:
PUMA_check1 = 'PUMA == 1300 | PUMA == 800 | PUMA == 105'
PUMA_check2 = 'PUMA == 105 | PUMA == 300 | PUMA == 1104'

PUMA_set1 = basedf.query(PUMA_check1)
PUMA_set2 = basedf.query(PUMA_check2)

# PUMA00_set1[['PUMA00']].drop_duplicates().head()

# PUMA00_set2[['PUMA00']].drop_duplicates().head()

# basedf.query("PUMA00==1104")

2) Use merge to perform and inner join between the two tables on SERIALNO (HINT use reset index() on each first to make the old index a column in each table)

In [108]:
PUMA_set1 = PUMA_set1.reset_index()
PUMA_set2 = PUMA_set2.reset_index()

In [110]:
PUMA_set1.head()

pd.merge(PUMA_set1, PUMA_set2, on='SERIALNO')

SERIALNO  PUMA_x  NP_x  BDSP_x  PUMA_y  NP_y  BDSP_y
0     2013000000051     105     0     1.0     105     0     1.0
1     2013000000175     105     1     3.0     105     1     3.0
2     2013000002431     105     1     2.0     105     1     2.0
3     2013000002854     105     7     9.0     105     7     9.0
4     2013000004409     105     2     3.0     105     2     3.0
5     2013000005563     105     1     1.0     105     1     1.0
6     2013000007495     105     1     1.0     105     1     1.0
7     2013000008635     105     1     3.0     105     1     3.0
8     2013000010068     105     1     1.0     105     1     1.0
9     2013000010527     105     2     2.0     105     2     2.0
10    2013000011881     105     1     NaN     105     1     NaN
11    2013000013696     105     2     2.0     105     2     2.0
12    2013000014177     105     2     3.0     105     2     3.0
13    2013000015644     105     2     3.0     105     2     3.0
14    2013000017719     105     1     0.0     105     1     0.0
15    2013000019989     105     2     2.0     105     2     2.0
16    2013000020604     105     1     2.0     105     1     2.0
17    2013000023410     105     2     2.0     105     2     2.0
18    2013000023853     105     2     3.0     105     2     3.0
19    2013000023872     105     3     9.0     105     3     9.0
20    2013000024856     105     2     4.0     105     2     4.0
21    2013000025590     105     2     3.0     105     2     3.0
22    2013000025803     105     8     3.0     105     8     3.0
23    2013000036411     105     2     4.0     105     2     4.0
24    2013000038366     105     1     NaN     105     1     NaN
25    2013000038457     105     2     3.0     105     2     3.0
26    2013000038480     105     2     4.0     105     2     4.0
27    2013000044774     105     1     2.0     105     1     2.0
28    2013000047000     105     1     3.0     105     1     3.0
29    2013000049008     105     1     1.0     105     1     1.0
...             ...     ...   ...     ...     ...   ...     ...
3904  2017001488553     105     1     NaN     105     1     NaN
3905  2017001491804     105     1     NaN     105     1     NaN
3906  2017001493533     105     1     NaN     105     1     NaN
3907  2017001496838     105     1     NaN     105     1     NaN
3908  2017001497668     105     1     NaN     105     1     NaN
3909  2017001497782     105     1     NaN     105     1     NaN
3910  2017001498242     105     1     NaN     105     1     NaN
3911  2017001498746     105     1     NaN     105     1     NaN
3912  2017001500114     105     1     NaN     105     1     NaN
3913  2017001500992     105     1     NaN     105     1     NaN
3914  2017001502002     105     1     NaN     105     1     NaN
3915  2017001502980     105     1     NaN     105     1     NaN
3916  2017001503716     105     1     NaN     105     1     NaN
3917  2017001505754     105     1     NaN     105     1     NaN
3918  2017001512737     105     1     NaN     105     1     NaN
3919  2017001515492     105     1     NaN     105     1     NaN
3920  2017001515742     105     1     NaN     105     1     NaN
3921  2017001517961     105     1     NaN     105     1     NaN
3922  2017001518627     105     1     NaN     105     1     NaN
3923  2017001519327     105     1     NaN     105     1     NaN
3924  2017001520820     105     1     NaN     105     1     NaN
3925  2017001521173     105     1     NaN     105     1     NaN
3926  2017001527521     105     1     NaN     105     1     NaN
3927  2017001529533     105     1     NaN     105     1     NaN
3928  2017001529622     105     1     NaN     105     1     NaN
3929  2017001530617     105     1     NaN     105     1     NaN
3930  2017001532343     105     1     NaN     105     1     NaN
3931  2017001534087     105     1     NaN     105     1     NaN
3932  2017001534807     105     1     NaN     105     1     NaN
3933  2017001536038     105     1     NaN     105     1     NaN

[3934 rows x 7 columns]

3) The resulting table will only have rows that appeared in both. Which PUMAs belong to the overlap?

PUMA == 105 were the rows that overlapped

### Problems 3

1) Write a map operation that scales NP as an exponent using the numpy function np.exp

2) Write a variation on the CROSS APPLY above without using apply

In [114]:
housingdf = pd.read_csv('psam_h17.csv', dtype={'PUMA':str})
basedf = pd.read_csv('psam_h17.csv', 
                      usecols=['SERIALNO', 'PUMA', 'ST',
                               'ADJHSG', 'ADJINC', 'WGTP', 'NP', 'TYPE', 'ACR',
                               'AGS', 'BATH', 'BDSP', 'BLD', 'BUS', 'CONP', 'ELEP',
                               'FS', 'FULP'])

In [117]:
housingdf['weightedNP'] = basedf['WGTP'] * basedf['NP']
g = housingdf.groupby(['PUMA'])
pumaAvgNPArray = g['weightedNP'].sum() / g['WGTP'].sum()
avgNPdf = pd.DataFrame(pumaAvgNPArray, columns=['avgNP']).reset_index()
avgNPdf.head()

PUMA     avgNP
0  00104  1.940075
1  00105  2.064938
2  00202  1.973556
3  00300  1.985774
4  00401  2.026103

## What is in a GeoDataBase
- A GeoDB has three essential components:
    - Spatial features (with a Datum and Projection information)
    - Attributes linked to spatial features
    - A means of transforming and linking by attribute data or spatial feature
- Pandas gives us a way of managing structured attribute data, what do we need to add in order to build a usable spatial analysis data structure

## PySAL for computational geometry
- PySAL provides a high level interface for shape objects and their transformation: pysal.lib.cg
- By using PySAL objects, we can read in and handle shape files (one of the typical formats for spatial data)
- We will examine some of the functionality that PySAL grants us in the context of an object column in a standard Pandas DataFrame
- note that PySAL is also an interface to a wide range of spatial statistical models and spatial econometric models that are not accessible from other python libraries

### Loeading our Pandas Data and our Shape File
- We can begin by loading a standard DataFrame from the csv data that we are interested in
- Note how we have handled the PUMA column
- Then we can load the shape file separately using the ps.lib.io.fileio.FileIO() function
- read() is a more fundamental IO method than the read csv() method and the file handle needs to be closed

In [1]:
import pandas as pd
import pysal as ps

housingdf = pd.read_csv('psam_h17.csv', dtype={'PUMA':str})
shp_path = 'tl_2018_17_puma10.shp'
f = ps.lib.io.fileio.FileIO(shp_path)
all_polygons = f.read()
f.close()

### Examining our spatial features
- The result of reading the shape file is a list of polygons

In [2]:
type(all_polygons)

list

In [3]:
type(all_polygons[0])

pysal.lib.cg.shapes.Polygon

In [4]:
all_polygons[0]

In [5]:
all_polygons[0].vertices

[(-87.721436, 41.734862),
 (-87.721417, 41.734863),
 (-87.72134799999999, 41.734863999999995),
 (-87.72113, 41.734866),
 (-87.721054, 41.734868),
 (-87.72064499999999, 41.734874),
 (-87.71941799999999, 41.734895),
 (-87.71901, 41.734902),
 (-87.718522, 41.734911),
 (-87.717168, 41.734938),
 (-87.71705999999999, 41.734939),
 (-87.716573, 41.734944999999996),
 (-87.71608599999999, 41.73495),
 (-87.715671, 41.734955),
 (-87.714629, 41.734984999999995),
 (-87.71457199999999, 41.734987),
 (-87.71414299999999, 41.734992999999996),
 (-87.713709, 41.734998),
 (-87.71300699999999, 41.735008),
 (-87.71240999999999, 41.73502),
 (-87.711978, 41.735029),
 (-87.711923, 41.735029),
 (-87.71176, 41.735032),
 (-87.71170599999999, 41.735034),
 (-87.711658, 41.735034999999996),
 (-87.71162199999999, 41.735036),
 (-87.710782, 41.735051999999996),
 (-87.709862, 41.735071),
 (-87.709245, 41.735079999999996),
 (-87.708012, 41.735099),
 (-87.70783999999999, 41.735102),
 (-87.707089, 41.735115),
 (-87.70609499

The polygon vertices are the longitude and latitude

PySAL Shapes include a substantial number of precomputed attributes and efficient methods

In [6]:
all_polygons[0].centroid

(-87.73227916862234, 41.68628900126081)

In [7]:
all_polygons[0].perimeter

0.7424176724174986

In [8]:
all_polygons[0].area

0.010349874158494149

In [9]:
ps.lib.cg.get_shared_segments(all_polygons[0], all_polygons[14])

The computed segments are the shared boundary of the two polygons.

### Where is the data associated with the shapes?
- In order to capture the attributes associated with the shapes themselves, we will need to read the associated .dbf file
- There are two reasons to do this:
    - First, the regions probably have some interesting additional data from the census
    - Second, we don’t have labels for these shapes so we can’t link them to the housingdf!

In [11]:
dfpoly = pd.DataFrame(all_polygons,columns=['polygon'])
dbf_path = 'tl_2018_17_puma10.dbf'
f2 = ps.lib.io.fileio.FileIO(dbf_path)
dbheader = f2.header
dbfile = f2.read()
f2.close()
pd.DataFrame(dbfile, columns=dbheader).head()

STATEFP10 PUMACE10  GEOID10  \
0        17    03411  1703411   
1        17    03107  1703107   
2        17    03700  1703700   
3        17    02000  1702000   
4        17    01602  1701602   

                                          NAMELSAD10 MTFCC10 FUNCSTAT10  \
0  Cook County (South Central)--Worth & Calumet T...   G6120          S   
1  Will County (Northeast)--Frankfort, Homer & Ne...   G6120          S   
2                     Kendall & Grundy Counties PUMA   G6120          S   
3                                 McLean County PUMA   G6120          S   
4  Menard, Logan, De Witt, Piatt, Moultrie, Shelb...   G6120          S   

      ALAND10  AWATER10   INTPTLAT10    INTPTLON10  
0    94129155   1576803  +41.7313634  -087.7167725  
1   243596923    478565  +41.5528956  -087.9168232  
2  1912412909  37262592  +41.4200983  -088.4339373  
3  3064559693   7853695  +40.4945594  -088.8445391  
4  9254202416  88234698  +39.7699432  -089.2258108

### Producing a complete data set
- The coordinates of the centroid of the first elements seem to agree
- Let’s try and combine them by pd.concat and then join to the csv file

In [14]:
geo_df1 = pd.concat([pd.DataFrame(dbfile, columns=dbheader),dfpoly],axis=1)
geo_df1.head()
pd.merge(geo_df1, housingdf,how='inner',left_on=['PUMACE10'],right_on=['PUMA'])

STATEFP10 PUMACE10  GEOID10  \
0             17    03411  1703411   
1             17    03411  1703411   
2             17    03411  1703411   
3             17    03411  1703411   
4             17    03411  1703411   
5             17    03411  1703411   
6             17    03411  1703411   
7             17    03411  1703411   
8             17    03411  1703411   
9             17    03411  1703411   
10            17    03411  1703411   
11            17    03411  1703411   
12            17    03411  1703411   
13            17    03411  1703411   
14            17    03411  1703411   
15            17    03411  1703411   
16            17    03411  1703411   
17            17    03411  1703411   
18            17    03411  1703411   
19            17    03411  1703411   
20            17    03411  1703411   
21            17    03411  1703411   
22            17    03411  1703411   
23            17    03411  1703411   
24            17    03411  1703411   
25            17    03411  1703411   
26            17    03411  1703411   
27            17    03411  1703411   
28            17    03411  1703411   
29            17    03411  1703411   
...          ...      ...      ...   
292898        17    03526  1703526   
292899        17    03526  1703526   
292900        17    03526  1703526   
292901        17    03526  1703526   
292902        17    03526  1703526   
292903        17    03526  1703526   
292904        17    03526  1703526   
292905        17    03526  1703526   
292906        17    03526  1703526   
292907        17    03526  1703526   
292908        17    03526  1703526   
292909        17    03526  1703526   
292910        17    03526  1703526   
292911        17    03526  1703526   
292912        17    03526  1703526   
292913        17    03526  1703526   
292914        17    03526  1703526   
292915        17    03526  1703526   
292916        17    03526  1703526   
292917        17    03526  1703526   
292918        17    03526  1703526   
292919        17    03526  1703526   
292920        17    03526  1703526   
292921        17    03526  1703526   
292922        17    03526  1703526   
292923        17    03526  1703526   
292924        17    03526  1703526   
292925        17    03526  1703526   
292926        17    03526  1703526   
292927        17    03526  1703526   

                                               NAMELSAD10 MTFCC10 FUNCSTAT10  \
0       Cook County (South Central)--Worth & Calumet T...   G6120          S   
1       Cook County (South Central)--Worth & Calumet T...   G6120          S   
2       Cook County (South Central)--Worth & Calumet T...   G6120          S   
3       Cook County (South Central)--Worth & Calumet T...   G6120          S   
4       Cook County (South Central)--Worth & Calumet T...   G6120          S   
5       Cook County (South Central)--Worth & Calumet T...   G6120          S   
6       Cook County (South Central)--Worth & Calumet T...   G6120          S   
7       Cook County (South Central)--Worth & Calumet T...   G6120          S   
8       Cook County (South Central)--Worth & Calumet T...   G6120          S   
9       Cook County (South Central)--Worth & Calumet T...   G6120          S   
10      Cook County (South Central)--Worth & Calumet T...   G6120          S   
11      Cook County (South Central)--Worth & Calumet T...   G6120          S   
12      Cook County (South Central)--Worth & Calumet T...   G6120          S   
13      Cook County (South Central)--Worth & Calumet T...   G6120          S   
14      Cook County (South Central)--Worth & Calumet T...   G6120          S   
15      Cook County (South Central)--Worth & Calumet T...   G6120          S   
16      Cook County (South Central)--Worth & Calumet T...   G6120          S   
17      Cook County (South Central)--Worth & Calumet T...   G6120          S   
18      Cook County (South Central)--Worth & Calumet T...   G6120          S   
19      Cook County (South Central)--Worth & Calum

### Is this enough to get things done?
- This is a useful structure (especially for doing complex spatial statistics)
- We can automate complex problems for filtering and producing attributes derived from spatial features
- We have two problems that remain unsolved by this: projection and fast spatial indexing

In [15]:
geo_df1['computed_area']=geo_df1['polygon'].map(lambda x: x.area)
geo_df1['total_listed_area']=geo_df1.ALAND10+geo_df1.AWATER10

### Problems

1) Following the slides, assemble a dataframe from the csv, dbf, and shp files name it geo df1

2) 2 Use the code from the “Is this enough...” slide to compute the area from the polygons and the area listed in the dbf